addons/l10n_in/models/account_move_reversal.py

```diff
@@ -0,0 +1,14 @@
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+from odoo import models
+
+class AccountMoveReversal(models.TransientModel):
+    _inherit = "account.move.reversal"
+
+    def reverse_moves(self, is_modify=False):
+        res = super().reverse_moves(is_modify)
+        if self.env.company.account_fiscal_country_id.code == "IN":
+            for move_line, new_move_line in zip(self.move_ids.line_ids, self.new_move_ids.line_ids):
+                if move_line.product_id and new_move_line.product_id and move_line.l10n_in_hsn_code != new_move_line.l10n_in_hsn_code:
+                    new_move_line.l10n_in_hsn_code = move_line.l10n_in_hsn_code
```

Comment:
Why is it declared as `copy=False`, and why don't we manage the copy mechanism in `copy_data`, maybe using the existence of `move_reverse_cancel` in the context

Reply:
Hello @william-andre ,
Thanks for review.

We declared copy=False in field l10n_in_hsn_code, because whenever we change hsn code in the invoice line it will be applied for that one time only then if we duplicate that invoice, hsn code must come from a product not from the previous invoice line.

Thank you.

Reply:
If you look at `intrastat_transaction_id`, it is set as `copy=True`, and the mechanism should be pretty similar...

Reply:
yes, but we can duplicate intrastat_transaction_id, and I discussed with the Indian PO we don't want to duplicate HSN code from the invoice line. and we want only to copy HSN code in the credit note, not in other cases. that's why I do this way.

Reply:
Can you please share the reasoning why you want top copy HSN code on Credit Note?
In EU, with Intrastat, I would prefer not to copy the Intrastat Commodity Code on Credit Notes for the following reason. Some intrastat commodity codes expire, are split in several other codes, are merged into other codes, on a yearly basis. Suppose an invoice on Year N and a Credit Note on Year N+1, if any of those accounts expire in the meatime, it would lead to errors and the users would have it hard to understand what to change, and how... on hidden fields, that could prevent him to submit his file. 

I was considering that not having on the credit note a code that is a copy of the one on the invoice would make it easier to understand. It's very technical to consider a Credit Note of an Invoice is first a copy of it and then an inversion. For the user, a credit note is cancelling an invoice. A credit note contains products (probably) and the commodity code comes from the product.

Reply:
Hello @ajjo-odoo ,
Can you please explain to him why we need to copy HSN code in credit note from invoice line?

Reply:
@tsb-odoo 

Reason for copy HSN code from Invoice in case of Credit note :

Normaly it's good practice to capture HSN from Product.
lets check with use case.
On 1st March 2024 , User sold LED TV -  HSN is 2244 of Rs. 1,00,000/- 
On dated 20th March 2024 government has declared that now HSN for LED TV would be 2254.
User got Sales return of the product on 31st march 2024.
Now which HSN code we should capture New or Old that's one of the challange we are facing and government guidelines are also silent on this.

based on our experience and after consulting several Chartered Accountants and expert based on their practice
We got following feedback.. to keep hsn mentioned in the invoice
1. If we capture New HSN which is updated on Product Master, our HSN summary will have negative value for the credit note in GSTR 1, so it's good to keep HSN mentioned in Invoice for the credit note raised
2. Earlier it was mandatory to provide reference of invoice in credit for GSTR 1, so if mapping is required good approach to keep details similar, so tax officer can have idea about Credit note raised against invoices.


we also got another approach what if edi is requires and old hsn is no more available.
But up to till date we have not seen such use case either from user or from expert.

We could search change of HSN but not the expiry of HSN.

Reply:
@tsb-odoo 
i believe we shall carry this approach until government comes up with some clarification.
thank You
cc - @jva-odoo @jasa-odoo

Reply:
> Why is it declared as `copy=False`, and why don't we manage the copy mechanism in `copy_data`, maybe using the existence of `move_reverse_cancel` in the context

Hello, @william-andre 
I tried different approaches. IMO, we shouldn't override the copy data method because it will duplicate HSN code in credit note and in duplicate invoices also. but that we didn't want. So if we want to use copy data, is the following way good enough or not?
```
for move_line in self.move_ids.invoice_line_ids:
   new_move_line = self.new_move_ids.invoice_line_ids.filtered(lambda l: l.product_id == move_line.product_id)
   if new_move_line and move_line.l10n_in_hsn_code != new_move_line.l10n_in_hsn_code:
        new_move_line.write(new_move_line.copy_data({'l10n_in_hsn_code': move_line.l10n_in_hsn_code})[0])
```

Thanks in advance :)

Reply:
> I tried different approaches. IMO, we shouldn't override the copy data method because it will duplicate HSN code in credit note and in duplicate invoices also.

Did you read my comment about the context?

Reply:
> > Why is it declared as `copy=False`, and why don't we manage the copy mechanism in `copy_data`, maybe using the existence of `move_reverse_cancel` in the context
> 
> Hello, @william-andre I tried different approaches. IMO, we shouldn't override the copy data method because it will duplicate HSN code in credit note and in duplicate invoices also. but that we didn't want. So if we want to use copy data, is the following way good enough or not?
> 
> ```
> for move_line in self.move_ids.invoice_line_ids:
>    new_move_line = self.new_move_ids.invoice_line_ids.filtered(lambda l: l.product_id == move_line.product_id)
>    if new_move_line and move_line.l10n_in_hsn_code != new_move_line.l10n_in_hsn_code:
>         new_move_line.write(new_move_line.copy_data({'l10n_in_hsn_code': move_line.l10n_in_hsn_code})[0])
> ```
> 
> Thanks in advance :)

We don't want to copy the HSN code on the credit notes either. 
It's a super-edge case and GSTR-1 accepts negative amounts on an HSN code, afaik.

Reply:
> > > Why is it declared as `copy=False`, and why don't we manage the copy mechanism in `copy_data`, maybe using the existence of `move_reverse_cancel` in the context
> > 
> > 
> > Hello, @william-andre I tried different approaches. IMO, we shouldn't override the copy data method because it will duplicate HSN code in credit note and in duplicate invoices also. but that we didn't want. So if we want to use copy data, is the following way good enough or not?
> > ```
> > for move_line in self.move_ids.invoice_line_ids:
> >    new_move_line = self.new_move_ids.invoice_line_ids.filtered(lambda l: l.product_id == move_line.product_id)
> >    if new_move_line and move_line.l10n_in_hsn_code != new_move_line.l10n_in_hsn_code:
> >         new_move_line.write(new_move_line.copy_data({'l10n_in_hsn_code': move_line.l10n_in_hsn_code})[0])
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Thanks in advance :)
> 
> We don't want to copy the HSN code on the credit notes either. It's a super-edge case and GSTR-1 accepts negative amounts on an HSN code, afaik.

@tsb-odoo  if you believe that such kind of use cases shall be handled manually than it's fine with.
As we have asked to Chartered Accountants and Government Q & A platform , 
if they would come up with the clarification that Credit - Debit note would always carry the same HSN of Invoice than we may re-open this PR.

CC- @jasa-odoo @jva-odoo